index.js

```diff
@@ -121,53 +113,46 @@ async function handleIncreaseScores (req, res, redis, signerAddresses, redlock,
   const scheduledRewardsDeltas = body.scores.map(score => {
     return (BigInt(score) * roundReward) / maxScore
   })
-  let updatedRewards
 
-  const lock = await redlock.lock('lock:rewards', 60_000)
+  const pgClient = await pgPool.connect()
   try {
-    const currentRewards = (await redis.hmget('rewards', ...body.participants)).map(amount => {
-      return BigInt(amount || '0')
-    })
-    updatedRewards = body.scores.map((_, i) => {
-      return currentRewards[i] + scheduledRewardsDeltas[i]
-    })
-
-    const tx = redis.multi()
-    tx.hset(
-      'rewards',
-      Object.fromEntries(body.participants.map((address, i) => ([
-        address,
-        String(updatedRewards[i])
-      ])))
-    )
-    for (let i = 0; i < body.participants.length; i++) {
-      addLogJSON(tx, {
-        timestamp,
-        address: body.participants[i],
-        score: body.scores[i],
-        scheduledRewardsDelta: String(scheduledRewardsDeltas[i])
-      })
-    }
-    await tx.exec()
+    await pgClient.query('BEGIN')
+    await pgClient.query(`
+      INSERT INTO scheduled_rewards (address, amount)
+      VALUES (UNNEST($1::TEXT[]), UNNEST($2::NUMERIC[]))
+      ON CONFLICT (address) DO UPDATE
+        SET amount = scheduled_rewards.amount + EXCLUDED.amount
+    `, [body.participants, scheduledRewardsDeltas])
+    await pgClient.query(`
+      INSERT INTO logs (address, score, scheduled_rewards_delta)
+      VALUES (UNNEST($1::TEXT[]), UNNEST($2::NUMERIC[]), UNNEST($3::NUMERIC[]))
+    `, [body.participants, body.scores, scheduledRewardsDeltas])
+    await pgClient.query(`
+      DELETE FROM logs WHERE timestamp < NOW() - INTERVAL '30 days'
+    `)
+    await pgClient.query('COMMIT')
+  } catch (err) {
+    await pgClient.query('ROLLBACK')
+    throw err
   } finally {
-    await lock.unlock()
+    pgClient.release()
   }
 
   logger.info(`Increased scheduled rewards of ${body.participants.length} participants`)
+
+  const { rows } = await pgPool.query(`
+    SELECT address, amount
+    FROM scheduled_rewards
+    WHERE address = ANY($1)  
+  `, [body.participants])
```

Comment:
I think you can avoid this extra query by modifying the first query `INSERT INTO scheduled_rewards` to return the inserted/updated scheduled rewards. 

Think of `INSERT ... RETURNING id` that's used to get the ID value generated by Postgres.

Reply:
I thought about this too, but above we're inside a transaction, with multiple inserts, so I don't know how to get its output 🤔